External/FEXCore/include/FEXCore/Utils/ThreadPoolAllocator.h

```diff
@@ -0,0 +1,507 @@
+            }
+          }
+
+          // If we didn't have an exact fit then use an unsized fit
+          if (BestFit == BuffersEnd) {
+            BestFit = UnsizedFit;
+          }
+
+          {
+            std::chrono::time_point<ClockType> LRUTime{};
+            ContainerType::iterator LastUsed = BuffersEnd;
+
+            // Walk all the allocations and find a buffer to erase
+            for (auto it = OwnedBuffers.begin(); it != OwnedBuffers.end(); ++it) {
+              // Ensure that the LRU value is past our duration threshold and isn't the one we are claiming
+              // Also only select a single memory region
+              if (it != BestFit) {
+                auto UsedTime = (*it)->LastUsed.load(std::memory_order_relaxed);
+                if ((Now - UsedTime) >= DURATION &&
+                    UsedTime > LRUTime) {
+                  LastUsed = it;
+                  LRUTime = UsedTime;
+                }
+              }
+            }
+
+            // If we found a buffer then free it
+            if (LastUsed != BuffersEnd) {
+              Free((*LastUsed)->Ptr, (*LastUsed)->Size);
+              delete *LastUsed;
+              OwnedBuffers.erase(LastUsed);
+            }
+          }
+
+          if (BestFit != OwnedBuffers.end()) {
+            MemoryBuffer *Buffer = *BestFit;
+            OwnedBuffers.erase(BestFit);
+            return NonOwnedBuffers.emplace(NonOwnedBuffers.end(), Buffer);
+          }
+        }
+
+        // Need to allocate a new buffer, couldn't fit
+        auto Data = Alloc(Size);
+        return NonOwnedBuffers.emplace(NonOwnedBuffers.end(), new MemoryBuffer{Data, Size, ClockType::now()});
+      }
+
+      void UnclaimBufferImpl(ContainerType::iterator Buffer) {
+        (*Buffer)->CurrentClientOwnedFlag = nullptr;
+        OwnedBuffers.emplace_back(*Buffer);
+        NonOwnedBuffers.erase(Buffer);
+      }
+
+      /**
+       * @brief List of buffers that this pool allocator itself owns
+       */
+      ContainerType OwnedBuffers;
+
+      /**
+       * @brief List of buffers that are client claimed
+       */
+      ContainerType NonOwnedBuffers;
+
+      /**
+       * @brief Mutex to ensure thread safety while shuffling buffers around and allocating
+       */
+      std::mutex AllocationMutex;
+
+    private:
+      /**
+       * @brief Allocates the buffer
+       *
+       * @param Size of the object to allocate
+       *
+       * @return pointer
+       */
+      virtual void *Alloc(size_t Size) = 0;
+      /**
+       * @brief Frees the buffer
+       *
+       * @param Ptr buffer pointer
+       * @param Size buffer size
+       */
+      virtual void Free(void* Ptr, size_t Size) = 0;
+  };
+
+  /**
+   * @brief Thread pool allocator that allocates and frees objects using malloc
+   */
+  class ThreadPoolAllocatorMalloc final : public IntrusiveThreadPoolAllocator {
+    public:
+      ThreadPoolAllocatorMalloc() = default;
+
+      virtual ~ThreadPoolAllocatorMalloc() {
```

Comment:
Since both dtors `ThreadPoolAllocatorMalloc`, `ThreadPoolAllocatorMMap` are the same, wouldn't it be nicer to have the dtor logic in `IntrusiveThreadPoolAllocator` and rely on the alloc/free overloads? Unless I'm missing something

Reply:
Can't
`Call to pure virtual member function 'Free' has undefined behavior; overrides of 'Free' in subclasses are not available in the destructor of 'IntrusivePooledAllocator'  ThreadPoolAllocator.h:323:7: note: 'Free' declared here [-Wcall-to-pure-virtual-from-ctor-dtor]`

Reply:
Doh, right, derived class is destructed at that point. How about moving them to a function so that we don't have two copies of the code around?

Reply:
Moved duplicated freeing logic to its own function. Resolved now.